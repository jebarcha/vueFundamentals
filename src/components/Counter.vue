<template>
    <h2>{{ customTitle }}  {{ start }}</h2>
    <p> {{ counter }} <sup>2</sup>= {{ squareCounted }}</p>
    <div>
        <!-- <button v-on:click="increase">+1</button> -->
        <button @click="increase">+1</button>
        <button @click="decrease">-1</button>
    </div>
</template>

<script>
export default {
    props: {
        title: String,
        start: {
            type: Number,
            default: 100,
            validator: function(value) {
                return (value > 100)
            }
            //required: true
        }
    },
    //props: ['title', 'start'],
    //name: 'loquesea',
    data() {
        return {
            counter: this.start //5
        }
    },
    methods: {
        getSquareValue() {
            console.log('getSquareValue');
            return this.counter * this.counter
        },
        increase() {
            //this.counter++
            this.counter = this.counter + 1
        },
        decrease() {
            this.counter--
        }
    },
    computed: {
        squareCounted() {
            console.log('computed squareCounted')
            return this.counter * this.counter
        },
        customTitle() {
            return this.title || 'Counter'
        }
    }
}
</script>

<style>

button {
    background-color: #64BB87;
    border-radius: 5px;
    border: 1px solid white;
    color: white;
    cursor: pointer;
    margin: 0 5px;
    padding: 5px 15px;
    transition: 0.3s ease-in-out;
}

button:hover {
    background-color: #5aa67b;
    transition: 0.3s ease-in-out;
}

</style>