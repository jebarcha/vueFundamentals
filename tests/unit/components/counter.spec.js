import { shallowMount } from '@vue/test-utils'
import Counter from '@/components/Counter'

describe('Counter Component', () => {
  
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(Counter)
    })

    test('should match with the snapshot', () => {
        expect(wrapper.html).toMatchSnapshot()
    })

    test('should h2 has default value', () => {
        expect(wrapper.find('h2').exists()).toBeTruthy();

        const h2Value = wrapper.find('h2').text()

        expect(h2Value).toBe('Counter')        
    })

    test('should counter contain default value of 100', () => {
        //const pTags = wrapper.findAll('p')
        //expect(pTags[1].text()).toBe('100')

        const value = wrapper.find('[data-testid="counter"]').text()
        expect(value).toBe('100')

    })

    test('should increment and decrement the counter', async() => {
        const [ increaseBtn, decreaseBtn ] = wrapper.findAll('button')

        await increaseBtn.trigger('click')
        await increaseBtn.trigger('click')
        await increaseBtn.trigger('click')
        await decreaseBtn.trigger('click')
        await decreaseBtn.trigger('click')

        const value = wrapper.find('[data-testid="counter"]').text()
        
        expect(value).toBe('101')
    })

    test('should establish the default value', () => {
        //console.log(wrapper.props())
        const { start } = wrapper.props()
        //const start = wrapper.props('start')
        //console.log(start, typeof start)
        
        const value = wrapper.find('[data-testid="counter"]').text()
        //console.log(value);

        expect( Number(value) ).toBe(start)
    })

    test('should show prop title', () => {
        const title = 'Hola Mundo!!!'
        const wrapper = shallowMount(Counter, {
            props: {
                title
            }
        })

        //console.log(wrapper.html());
        expect(wrapper.find('h2').text()).toBe(title)
    })

})