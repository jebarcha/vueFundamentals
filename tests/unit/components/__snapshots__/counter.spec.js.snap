// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component should match with the snapshot 1`] = `[Function]`;
